[WIP] feat: harden post type usage and establish relationships between listings and posts/pages

dkoo opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

This one is kind of big! Does a few things to lay the groundwork for establishing cross-data-type relationships:

1. Restricts the use of Business Listing patterns to the Place listing post type. This establishes Places as the de facto business data type, and Marketplace listings as more one-off classifieds which can be "owned" by one or more Places.
2. Creates shadow taxonomies for each listing post type. These taxonomies are automatically created, updated, or deleted when posts of the corresponding listing types are created, renamed, or unpublished, respectively.
3. Shadow taxonomies can be assigned to various post types. This lets us create associations between posts of different types—for example, Places can be assigned to Marketplace listings and Events. Listings can also be associated with other listings of the same type, which should let us show "related listings" more easily. This closes #33. 
4. Each shadow taxonomy can be assigned to specific listings of other types, and all can be assigned to pages and posts. Closes #34.
5. When viewing a post, page, or listing that's been assigned listing shadow terms, we can automatically show a blurb and link for those associated listings. I've implemented a test version of this using the existing `.author-bio` styles from the Newspack theme, but this might need more design consideration (cc @thomasguillot).
6. Adds a couple of test block patterns for Marketplace listings (in the form of real estate listings). These are definitely placeholder patterns that I created mainly for testing purposes, so we'll want to revisit in another PR (cc @thomasguillot).
7. Adds a custom "Price" block that I figured would be useful for Marketplace listings (e.g. classifieds and real estate listings). This block editor is a fairly simple number input that also lets you select from a list of WooCommerce currencies (if available—defaults to the site's default currency, and falls back to USD if WooCommerce is not available) and then formats the number shown on the front-end with the corresponding currency symbol and internationalization based on the editor's browser settings—so if you choose Pound Sterling and you as an admin have a browser set to `en-GB`, it should show the number 10000 as `£10,000.00`. If you choose Euro and your browser is set to `fr-FR`, it should show `10 000,00 EUR`. I based the formatting on the admin's browser settings because I figured that the price should be shown in the format familiar to the person creating the listing and choosing the currency. This block is also synced to a single meta field for listing CPTs, which automatically inherits the value from first instance of a Price block. This should be useful for future search and filtering features.

### How to test the changes in this Pull Request:

#### Testing shadow taxonomies

1. After checking out this branch, edit a post or page—you should now see sidebar panels for each listing post type alongside categories and tags. Also confirm that terms are automatically created for pre-existing published listings (so editors who are already running listings won't need to go back and update or recreate those to get the new terms).

<img width="282" alt="Screen Shot 2021-03-31 at 4 31 51 PM" src="https://user-images.githubusercontent.com/2230142/113227444-ef4ffb00-924f-11eb-8311-615f01c82eb9.png">

2. Edit listings of each type and confirm that you see sidebar panels for the listing post types designated as being assignable in `class-newspack-listings-taxonomies.php`. For example, you should be able to assign Events to other Events and Generic listings, Generic listings only to other Generic listings, Marketplace listings to Generic listings and other Marketplace listings, and Places to all listing types.
3. Assign some listings to posts, pages, and other listings and view the front-end. 
4. You should see blurbs for each assigned listing at the bottom of the content (see below):

**Note:** I'm not sure the language ("Listed by", "Related listings") is quite right, but I'm also not sure what these headings should say. It seems particularly weird to show "Listed by" when some of the blurbs are of the same type (e.g., showing an Event blurb on another Event), but I'm also not sure what we should show in the case of related listings of the same type. The blurbs also visually clash with author bios and sponsor blurbs, if those are enabled. Open to suggestions.

For example, on a post or page:
<img width="803" alt="Screen Shot 2021-03-31 at 4 31 34 PM" src="https://user-images.githubusercontent.com/2230142/113227968-165afc80-9251-11eb-9b46-3ebb2f685071.png">

On another listing:
<img width="809" alt="Screen Shot 2021-03-31 at 4 25 36 PM" src="https://user-images.githubusercontent.com/2230142/113227950-0ba06780-9251-11eb-93d3-885e4f405aa0.png">

#### Testing shadow taxonomy updating/deleting

1. Publish a new listing of any type.
2. Edit a post, page, or another listing that is able to receive that listing type as a shadow term.
3. Confirm that the new listing is searchable/selectable as a term in the editor sidebar.
4. Revert the listing to a draft.
5. Refresh the editor for the post/page/other listing and confirm that the listing term is no longer selectable or searchable.
6. Publish a new post of any type other than a listing and confirm that no new term is created under any of the listing shadow taxonomies.

#### Testing Price block

1. Create or edit a Marketplace listing.
2. Confirm that a new block called "Price" is now insertable through the Block Inserter. Conversely, confirm that the Price block is not available in any other post type.

<img width="719" alt="Screen Shot 2021-03-31 at 6 48 48 PM" src="https://user-images.githubusercontent.com/2230142/113228729-bf562700-9252-11eb-926f-2adb13b7bc1c.png">

3. Test the block functionality:
  - Enter numbers in the input field. Confirm that you can only enter numbers greater than or equal to 0.
  - Confirm that the Currency field defaults to the currency set for your site in WooCommerce settings (which is set during Newspack onboarding). If you want to go above and beyond, you can install Listings on a fresh site without WooCommerce and confirm that a.) the Currency field is hidden, and b.) formatted prices are only shown in US dollars.
  - Confirm that when unfocusing the input field, the number is automatically formatted with the selected currency and decimals (if "Show decimals" is off, then the number is rounded to the nearest integer, if on, then it's shown to the nearest hundredth). Confirm that the same is true on the front-end.
  - Choose various currencies and test the symbols and formatting in both the editor and front-end.
4. Change your browser language to something that will result in different number formatting than your usual language. For example, my Firefox is usually set to `en-US`, so I set it to `fr-FR` and restarted the browser app.
5. Edit the Marketplace listing with your Price block and confirm that the number shown is formatted according to your browser's language setting. For example, before setting my Firefox to `fr-FR`, the number 500000 in EUR is shown as `€500,000.00`. After changing to French, it's shown as `500 000,00 €`.
6. Update the post and view on the front-end, and confirm the same. Verify that the format shown on the front-end matches the formatting last saved, _not_ the reader's browser language.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

adekbadek commented on includes/class-newspack-listings-taxonomies.php:
```diff
@@ -0,0 +1,451 @@
+
+/**
+ * Blocks class.
+ * Sets up custom blocks for listings.
```
Not really ;)

dkoo replied:
Fixed in a70feef

adekbadek commented on includes/class-newspack-listings-taxonomies.php:
```diff
@@ -0,0 +1,451 @@
+	}
+
+	/**
+	 * When a Place changes status, add/update the shadow term if the status is `publish`, otherwise delete it.
```
Why is only one of the four listing types handled here? After reverting a generic listing to draft, it was still available in the taxonomy sidebar on a post.

dkoo replied:
This was an oversight (since I started off with just one post type for testing). Should be fixed in a6f0195.

adekbadek replied:
Not sure if that worked, I've found this issue:
1. Create a Place
2. Assign it to a Marketplace
3. Rename the place
4. Edit the Marketplace - there is now an additional Place taxonomy item - both old and new name are listed are assignable taxonomy options

adekbadek replied:
Also, the comment still mentions "Place" ;)

adekbadek reviewed (changes requested):
The core functionality, as well as the new block, work well. There are some issues though - apart from the ones in separate comment:

1. The list of the assignable listings should not contain the edited listing. 
2. In the taxonomy panel in the sidebar, after clicking "Add new <listing-type>", the text above the input says: "New Category Name". I'd expect it to say "New <listing-type>". After filling out this new listing name, confirming, and publishing, the listing is not created. I think the "Add new…" should rather link to editor creating a new listing.

dkoo commented on includes/class-newspack-listings-taxonomies.php:
```diff
@@ -0,0 +1,451 @@
+				$listing_url     = get_permalink( $related_listing_id );
+				$featured_image  = get_the_post_thumbnail( $related_listing_id, 'thumbnail', [ 'class' => 'avatar' ] );
+
+				$content .= '<div class="author-bio sponsor-bio newspack-listings__related-listing">'; // <a href="'. $listing_url . '">';
```
The `.sponsor-bio` styles are actually set in the Newspack theme, and it makes the image a tiny bit larger than the regular `.author-bio` styles. It also removes the border radius to make the avatars appear circular (which shouldn't apply to these listings). I used these existing theme styles so that these elements would hopefully inherit styles from the child themes as well as theme customizations such as colors, but I'm open to rendering markup and styles specific to the plugin if you think that would be a better solution.

dkoo commented on includes/class-newspack-listings-taxonomies.php:
```diff
@@ -0,0 +1,451 @@
+			return false;
+		}
+
+		$post_id     = $post_id;
```
Whoops, fixed in f2d554c

dkoo commented on src/blocks/price/edit.js:
```diff
@@ -0,0 +1,88 @@
+								.map( _currency => {
+									return {
+										value: _currency,
+										label: `${ currencies[ _currency ] } (${ _currency })`,
```
Fixed in 621c497

dkoo commented on src/blocks/price/view.php:
```diff
@@ -0,0 +1,43 @@
+		true
+	);
+
+	// Register Evend Dates block.
```
Whoops, missed this one because I was looking for "event". 😓  Fixed in a70feef

dkoo commented on src/blocks/price/edit.js:
```diff
@@ -0,0 +1,88 @@
+				currency: currency || defaultCurrency,
+				minimumFractionDigits: showDecimals ? 2 : 0,
+				maximumFractionDigits: showDecimals ? 2 : 0,
+			} ).format( price ),
```
Nice catch—fixed in 9ae4141

dkoo commented on includes/class-newspack-listings-block-patterns.php:
```diff
@@ -146,7 +132,33 @@ public static function register_block_patterns() {
+				],
+			],
+			'real_estate_2' => [
+				'post_types' => [ Core::NEWSPACK_LISTINGS_POST_TYPES['marketplace'] ],
```
Strange, this was working for me last week but no longer. I find that registering the patterns on `init` (instead of `admin_init`) seem to fix it for me, what about you? d27b19a

adekbadek replied:
It does!

dkoo commented:
> The list of the assignable listings should not contain the edited listing.

I couldn't figure out a more elegant solution than a50068a, do you have any better suggestions?

> In the taxonomy panel in the sidebar, after clicking "Add new ", the text above the input says: "New Category Name". I'd expect it to say "New ". After filling out this new listing name, confirming, and publishing, the listing is not created. I think the "Add new…" should rather link to editor creating a new listing.

Good idea—updated in ce6aa70.

adekbadek commented on includes/class-newspack-listings-taxonomies.php:
```diff
@@ -0,0 +1,512 @@
+		}
+
+		$post_id     = $post->ID;
+		$new_term_id = $new_term->term_id;
```
Looks like `$new_term` is an array, not object:

```
PHP Notice:  Trying to get property 'term_id' of non-object
```

adekbadek commented on src/editor/shadow-taxonomies/index.js:
```diff
@@ -0,0 +1,40 @@
+		const modifiedProps = { ...props, hasCreateAction: false };
+		return (
+			<div className="newspack-listings__shadow-taxonomy-control">
+				<OriginalComponent { ...modifiedProps } />
```
Enhancement: it might take a moment for the taxonomies to load - maybe a spinner could be shown in the meantime?

adekbadek reviewed (changes requested):
Sorry, something made GH submit the review 👆 before I wanted it to 🤷‍♀️

> The .sponsor-bio styles are actually set in the Newspack theme, 

It looks ok now, for some reason (perhaps theme was out of date locally) the image was extremely small before.

> I couldn't figure out a more elegant solution than a50068a, do you have any better suggestions?

Hm, this looks pretty hacky. Maybe there's a way to do it in the editor? The listing linking could even be a separate custom panel, which would make the sidebar more readable.

dkoo commented:
> Not sure if that worked, I've found this issue:
> 1. Create a Place
> 2. Assign it to a Marketplace
> 3. Rename the place
> 4. Edit the Marketplace - there is now an additional Place taxonomy item - both old and new name are listed are assignable taxonomy options

Ah, I see the problem. We were looking up shadow terms by post title, so if you update the post title, it thinks a term doesn't exist, so we go and create it.

In 4f374be I added a second check to look for shadow terms by slug. So now the term is only created if we can't find a shadow term by both post title and post slug.

As part of this testing I also saw a lot of "orphaned" terms getting created—terms that aren't connected to a post anymore. In the same commit I added an admin handler to look for any terms that don't have a corresponding post and delete them.

dkoo commented:
> Maybe there's a way to do it in the editor? The listing linking could even be a separate custom panel, which would make the sidebar more readable.

Are you thinking a single custom sidebar panel to hold all of these listing types? Maybe an autocomplete field for each type that can be associated with the current post?

adekbadek commented:
Yes, that's way better 🙌

dkoo commented:
Thanks for the feedback and suggestions. I'm going to put this PR on hold for a bit while I spend some time building out this custom UI.

dkoo commented:
@adekbadek After some thought I realized that listings don't just have a single type of relationship with each other and other posts/pages—it's definitely more of a parent/child relationship. I also struggled to think of a real use-case for applying CPTs to others of the same type (e.g. a Marketplace Listing owning another Marketplace Listing), so this is now no longer part of the UI. I still use the shadow taxonomy to apply only the shadow term for each listing CPT to itself, so we can query those and create missing terms and delete orphaned terms, but now this is the only purpose for being able to apply shadow terms to their own post types.

I split the UI into two panels to reflect these changes:

<img width="285" alt="Screen Shot 2021-04-16 at 4 19 35 PM" src="https://user-images.githubusercontent.com/2230142/115089567-b78fb700-9ecf-11eb-80ce-1e7c1a70a9e4.png">

We don't currently show the child listings on the parent's front-end, but I would like to add a grid-based component to show these. For example, if you have a Place representing a real estate firm which is applied as a parent to several Marketplace Listings that represent its property listings, you should optionally see a grid or list of those property listings somewhere on the real estate firm's page. We can maybe tackle this in a separate PR since this one is getting big.

Some other notes:

* There are two new global options (and per-post overrides) to hide parent and child listings that let admins choose whether these are displayed on the front-end. Both options are disabled by default, meaning parents and children will be shown on the front-end (once implemented).

* I changed "Add new (post type)" to a "View All" link instead, because I felt this is more useful if you have a lot of listings and can't remember the exact title to search for.

* When editing a listing CPT, the parent/child autocomplete fields are shown for only the post types that apply to the currently edited listing. So for Places, for example, they can own listings of any type, pages, and posts, but no other listing can own Places, so you'll only see the "Child Listings" panel for those listing types. Generic Listings can't own any posts other than regular posts or pages, so you'll only see those post types in its Child Listings panel. Any listing type can be applied to a regular post or page, but since posts and pages can't be applied as a parent to listings, that's a simpler relationship and the UI accordingly calls this "Related Listings" when editing a regular post or page.

* Another challenge with the UI I ran into was that when applying parent listings to a post, these are treated as taxonomy terms in the editor, so they're not actually saved until you "Save" or "Update" the post. However, applying child posts uses a custom REST endpoint that looks up each post in the autocomplete field and applies the current post's shadow term to all of them (or removes it from those that are removed from the UI). Because this happens separately from the regular post editor functionality, these updates happen immediately. I added some messaging to notify the user when the update happens, but it does feel somewhat strange to have two similar-looking UIs behave differently. We could potentially store child post IDs in a meta field and then update each child post on `insert_post`, but I worry that would overcomplicate an already pretty complicated thing, and I worry about the post meta getting out of sync with which posts actually have the terms (e.g. when terms get deleted). Another potential solution would be to use a custom REST endpoint to assign parents as well, but that seems overkill since the standard editor functionality works just fine for that.

I welcome your feedback, suggestions, and testing for this UI. Also paging @thomasguillot since this is UI-heavy work, in case you have any design-related feedback or want to take a crack at designing something different.

This is clearly a work-in-progress, so I'm adding a WIP label to this PR to reflect that.

adekbadek commented on includes/class-newspack-listings-api.php:
```diff
@@ -78,6 +79,71 @@ public static function register_routes() {
+			]
+		);
+
+		// GET listings taxonomy terms by name search term.
```
This should be updated - it's a POST request.

dkoo replied:
Ah yes, updated in a3978bd

adekbadek commented on src/editor/shadow-taxonomies/index.js:
```diff
@@ -0,0 +1,274 @@
+					{ ! newspack_listings_hide_children &&
+						filteredChildPostTypes.map( childPostType => {
+							return (
+								<TaxonomySearch
```
When editing a marketplace, I see posts & pages field. After inputing text, nothing is found and the request returns `404`. 

![image](https://user-images.githubusercontent.com/7383192/115691983-d8885a00-a35e-11eb-96c8-7c7c802f2502.png)

dkoo replied:
Fixed in 527fe4d—forgot that the core `posts` and `pages` endpoint have the plural `s`.

Also in the same commit: since only published listings get shadow terms, we now hide the Child Listings UI if the post you're editing isn't published. Also added a `validate_post` method to the taxonomies class to make it a bit easier to check whether a post should have a shadow term.

adekbadek replied:
The request now works, but I see an error in the logs: 
```
Trying to get property 'term_id' of non-object
```
Originating from here: 

https://github.com/Automattic/newspack-listings/blob/527fe4d35e4b1df818a4c41a8e10d656c8afa8bf/includes/class-newspack-listings-taxonomies.php#L578

Where `$shadow_term` is false, when I'm linking a just-published Event to a post or a page.

dkoo replied:
I made some logical errors in the `validate_post` method in the prior commit. Should be fixed in 4b43a7c.

adekbadek reviewed (changes requested):
> When editing a listing CPT, the parent/child autocomplete fields are shown for only the post types that apply to the currently edited listing. So for Places, for example, they can own listings of any type, pages, and posts, but no other listing can own Places […]

This is getting complex - adding some tests would come a long way in demystifying those relations.

dkoo commented:
> This is getting complex - adding some tests would come a long way in demystifying those relations.

It is getting complex, isn't it? I agree that this plugin is in need of some test coverage, but I feel like adding them here would make this PR even bigger and harder to review. Should we consider adding a test suite in a separate PR after this lands? Since we're not release Phase 2 features until all are complete, we'd have a chance to catch any errors with this PR again once the tests are implemented.

dkoo commented on src/editor/index.js:
```diff
@@ -30,12 +32,26 @@ if ( isListing() ) {
 	// Register Event Dates block if we're editing an Event.
-	if ( isListing( post_types.event.name ) ) {
+	if ( isListing( postTypes.event.name ) ) {
 		registerEventDatesBlock();
```
I'm not able to replicate this. Just in case, could you `npm ci` or `npm install` and try again after rebuilding?

adekbadek replied:
Whoops, can't replicate anymore. I was thinking there might be an issue with block registration, but it must've been something else.

dkoo commented on src/editor/index.js:
```diff
@@ -30,12 +32,26 @@ if ( isListing() ) {
+	}
+
+	// Filter taxonomy UI for listing shadow taxonomies.
+	// addFilter( 'editor.PostTaxonomyType', 'newspack-listings', ShadowTaxonomies );
```
Nope, removed in df7843f

adekbadek reviewed (changes requested):
(commented in existing thread)

adekbadek commented:
> Should we consider adding a test suite in a separate PR after this lands? Since we're not releasing Phase 2 features until all are complete, we'd have a chance to catch any errors with this PR again once the tests are implemented.

Since this is not merged to `master`, the tests can be added later, but I'd be wary of merging Phase 2 as a single huge PR. This would introduce a lot of potential of errors.

dkoo commented:
>  I'd be wary of merging Phase 2 as a single huge PR. This would introduce a lot of potential of errors.

Definitely something I've thought about, but the features outlined for a Phase 2 release don't make much sense to roll out individually. The approach we've taken in reviewing and merging smaller PRs against a clone of `master` (while minimizing updates to `master` in the meantime) should hopefully minimize the errors when we finally merge that branch.

adekbadek commented on includes/class-newspack-listings-taxonomies.php:
```diff
@@ -0,0 +1,664 @@
+	public static function validate_post( $post ) {
+		$is_valid_post = true;
+
+		// Post must be published.
```
```suggestion
		// Post must not be published.
```
Since the check below was updated. But that's weird, because I'd expect a published post to be linkable to a listing. Is that check below intended?

dkoo replied:
It's somewhat confusing (I even confused myself!) because `validate_post` returns `true` if the listing is "valid" (e.g. it should have a shadow term), and `false` if it's not valid. So the check should be: if the post is **not** published, return `false`, which is what the change in 4b43a7c reflects.

adekbadek replied:
Then the word "validate" is perhaps not the best - how about `should_update_shadowterm` as the method name?
Also, the code comment should still be updated ;)

dkoo replied:
`should_update_shadow_term` does make the function's purpose clearer. Updated in a136170.

dkoo commented:
@thomasguillot Since your redesigned UI won't affect the back-end portion of this PR, and this isn't getting released until all phase 2 PRs are approved, I'm going to merge this and handle the new UI in a separate PR.
